src/oemof/network/network/hierachical_nodes.py

```diff
@@ -0,0 +1,136 @@
+# -*- coding: utf-8 -*-
+"""This package contains the differnt types of hierachical Nodes for
+modelling an energy system graph.
+
+SPDX-FileCopyrightText: Stephan Günther <>
+SPDX-FileCopyrightText: Uwe Krien <krien@uni-bremen.de>
+SPDX-FileCopyrightText: Simon Hilpert <>
+SPDX-FileCopyrightText: Cord Kaldemeyer <>
+SPDX-FileCopyrightText: Patrik Schönfeldt <patrik.schoenfeldt@dlr.de>
+
+SPDX-License-Identifier: MIT
+"""
+
+from collections import deque
+
+from .helpers import HierachicalLabel
+from .nodes import Node
+
+
+class AtomicNode(Node):
+    def __init__(
+        self,
+        label,
+        *,
+        inputs=None,
+        outputs=None,
+        parent_node=None,
+        custom_properties=None,
+    ):
+        if not isinstance(label, HierachicalLabel):
+            label = HierachicalLabel(label=label, parent=parent_node)
+        super().__init__(
+            label=label,
+            inputs=inputs,
+            outputs=outputs,
+            custom_properties=custom_properties,
+        )
+
+
+class SubNetwork(Node):
+    def __init__(
+        self,
+        label,
+        *,
+        parent_node=None,
+        custom_properties=None,
+    ):
+        if not isinstance(label, HierachicalLabel):
+            label = HierachicalLabel(label=label, parent=parent_node)
+        super().__init__(label=label, custom_properties=custom_properties)
+
+        self.subnodes = []
+
+        # TODO: Try to avoid this local `import`.
+        from ..energy_system import EnergySystem
+
+        EnergySystem.signals[EnergySystem.add].connect(
+            self.add_subnodes, sender=self
+        )
+
+    def append_subnodes(self, *args):
+        for sub_component in args:
+            self.subnodes.append(sub_component)
+
+    def add_subnodes(self, node, **kwargs):
+        """Add subnodes to an EnergySystem.
+
+        This is meant to be used as an event callback that is called when this
+        node is added to an EnergySystem, to add the child nodes to the
+        EnergySystem, too.
+        """
+        # TODO:
+        #    Explain why the `node` argument is necessary.
+        assert self is node
```

Comment:
```suggestion
```

Reply:
Either we really check
```suggestion
        if self != node:
            raise ValueError("Call needs to be obj.add_cubnodes(obj)".)
```
or let it be.